src/wasm/wasm-binary.cpp

```diff
@@ -945,6 +945,15 @@ void WasmBinaryBuilder::read() {
 
     auto oldPos = pos;
 
+    // almost no sections can appear more than once
+    if (seenSections.count(BinaryConsts::Section(sectionCode)) > 0) {
+      if (sectionCode != BinaryConsts::Section::User && sectionCode != BinaryConsts::Section::Code) {
+        throw ParseException("section seen more than once: " + std::to_string(sectionCode));
+      }
+    } else {
+      seenSections.insert(BinaryConsts::Section(sectionCode));
```

Comment:
The main ones do, but some others don't like user sections? I guess those can also appear twice too, so this happens to be ok...

Anyhow, we should eventually check the order regardless since it's correct, as you say. Meanwhile #995 does @dschuff's optimizations.